src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -0,0 +1,678 @@
+                res.Resolution is ResolvedTypeKind.TypeParameter tp && typeParam.Item1.Equals(tp.Item.Origin) && typeParam.Item2.Value != tp.Item.TypeName.Value;
+
+            foreach (var resolutionDictionary in resolutionDictionaries)
+            {
+                // Contains a lookup of all the keys in the combined resolutions whose value needs to be updated
+                // if a certain type parameter is resolved by the currently processed dictionary.
+                var mayBeReplaced = combinedBuilder
+                    .Where(kv => kv.Value.Resolution.IsTypeParameter)
+                    .ToLookup(
+                        // ToDo: Check for composite types that contain type parameter types.
+                        kv => AsTypeResolutionKey(((ResolvedTypeKind.TypeParameter)kv.Value.Resolution).Item),
+                        entry => entry.Key);
+
+                // Do any replacements for type parameters that may be replaced with values in the current dictionary.
+                // This needs to be done first to cover an edge case.
+                foreach (var (typeParam, paramRes) in resolutionDictionary.Where(entry => mayBeReplaced.Contains(entry.Key)))
+                {
+                    // Get all the parameters whose value is dependent on the current resolution's type parameter,
+                    // and update their values with this resolution's value.
+                    foreach (var keyInCombined in mayBeReplaced[typeParam])
+                    {
+                        // Check that we are not constricting a type parameter to another type parameter of the same callable.
+                        success = success && !IsConstrictiveResolution(keyInCombined, paramRes);
+                        combinedBuilder[keyInCombined] = paramRes;
+                    }
+                }
+
+                // Add all resolutions to the current dictionary.
+                foreach (var (typeParam, paramRes) in resolutionDictionary)
+                {
+                    // Check that we are not constricting a type parameter to another type parameter of the same callable.
+                    success = success && !IsConstrictiveResolution(typeParam, paramRes);
+                    // Check that there is no conflicting resolution already defined.
+                    var conflictingResolutionExists = combinedBuilder.TryGetValue(typeParam, out var current)
+                        && !current.Equals(paramRes) && !IsSelfResolution(typeParam, current);
+                    success = success && !conflictingResolutionExists;
+                    combinedBuilder[typeParam] = paramRes;
+                }
+            }
+
+            combined = combinedBuilder.ToImmutable();
+            return success;
+        }
+    }
+
+    /// <summary>
+    /// Contains the information that exists on edges in a call graph.
+    /// The ParamResolutions are expected to be non-null and have all of their position information removed.
```

Comment:
Since the class constructor enforces that, I believe what you wanted to document was that other code can/should indeed rely on the ParamResolutions being non-null, and having position information stripped:
```suggestion
    /// The ParamResolutions are non-null and have all of their position information removed.
```

Reply:
done

Reply:
I believe you forgot to push a change.

Reply:
Sorry this got resolved on the wrong branch before. It has been cherry-picked into this branch now.